pkg/volume/fc/fc.go

```diff
@@ -0,0 +1,197 @@
+	return plugin.newBuilderInternal(spec, pod.UID, &FCUtil{}, mounter)
+}
+
+func (plugin *fcPlugin) newBuilderInternal(spec *volume.Spec, podUID types.UID, manager diskManager, mounter mount.Interface) (volume.Builder, error) {
+	// fc volumes used directly in a pod have a ReadOnly flag set by the pod author.
+	// fc volumes used as a PersistentVolume gets the ReadOnly flag indirectly through the persistent-claim volume used to mount the PV
+	var readOnly bool
+	var fc *api.FCVolumeSource
+	if spec.Volume != nil && spec.Volume.FC != nil {
+		fc = spec.Volume.FC
+		readOnly = fc.ReadOnly
+	} else {
+		fc = spec.PersistentVolume.Spec.FC
+		readOnly = spec.ReadOnly
+	}
+
+	lun := strconv.Itoa(fc.Lun)
+
+	return &fcDiskBuilder{
+		fcDisk: &fcDisk{
+			podUID:  podUID,
+			volName: spec.Name(),
+			wwns:    fc.TargetWWNs,
+			lun:     lun,
+			manager: manager,
+			mounter: &mount.SafeFormatAndMount{mounter, exec.New()},
+			plugin:  plugin},
+		fsType:   fc.FSType,
+		readOnly: readOnly,
+	}, nil
+}
+
+func (plugin *fcPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newCleanerInternal(volName, podUID, &FCUtil{}, mounter)
+}
+
+func (plugin *fcPlugin) newCleanerInternal(volName string, podUID types.UID, manager diskManager, mounter mount.Interface) (volume.Cleaner, error) {
+	return &fcDiskCleaner{&fcDisk{
+		podUID:  podUID,
+		volName: volName,
+		manager: manager,
+		mounter: mounter,
+		plugin:  plugin,
+	}}, nil
+}
+
+func (plugin *fcPlugin) execCommand(command string, args []string) ([]byte, error) {
+	cmd := plugin.exe.Command(command, args...)
+	return cmd.CombinedOutput()
+}
+
+type fcDisk struct {
+	volName string
+	podUID  types.UID
+	portal  string
+	wwns    []string
+	lun     string
+	plugin  *fcPlugin
+	mounter mount.Interface
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager diskManager
+}
+
+func (fc *fcDisk) GetPath() string {
+	name := fcPluginName
+	// safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
+	return fc.plugin.host.GetPodVolumeDir(fc.podUID, util.EscapeQualifiedNameForDisk(name), fc.volName)
+}
+
+type fcDiskBuilder struct {
+	*fcDisk
+	readOnly bool
+	fsType   string
+}
+
+var _ volume.Builder = &fcDiskBuilder{}
+
+func (b *fcDiskBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+func (b *fcDiskBuilder) SetUpAt(dir string) error {
+	// diskSetUp checks mountpoints and prevent repeated calls
+	err := diskSetUp(b.manager, *b, dir, b.mounter)
+	if err != nil {
+		glog.Errorf("fc: failed to setup")
+		return err
+	}
+	globalPDPath := b.manager.MakeGlobalPDName(*b.fcDisk)
+	var options []string
+	if b.readOnly {
+		options = []string{"remount", "ro"}
+	} else {
+		options = []string{"remount", "rw"}
+	}
+	return b.mounter.Mount(globalPDPath, dir, "", options)
```

Comment:
this is fixed

----- Original Message -----
From: "Saad Ali" <notifications@github.com>
To: "kubernetes/kubernetes" <kubernetes@noreply.github.com>
Cc: "Huamin Chen" <hchen@redhat.com>
Sent: Monday, September 14, 2015 2:44:24 PM
Subject: Re: [kubernetes] support fibre channel volume (#12599)

> +
> +func (b *fcDiskBuilder) SetUpAt(dir string) error {
> +	// diskSetUp checks mountpoints and prevent repeated calls
> +	err := diskSetUp(b.manager, *b, dir, b.mounter)
> +	if err != nil {
> +		glog.Errorf("fc: failed to setup")
> +		return err
> +	}
> +	globalPDPath := b.manager.MakeGlobalPDName(*b.fcDisk)
> +	var options []string
> +	if b.readOnly {
> +		options = []string{"remount", "ro"}
> +	} else {
> +		options = []string{"remount", "rw"}
> +	}
> +	return b.mounter.Mount(globalPDPath, dir, "", options)

What bug is this trying to work around? #10416 is resolved. Why is that no longer needed and this is?

Just want to make sure that we're not adding unnecessary code, but, also make sure we're not removing a legitimate work around.

---
Reply to this email directly or view it on GitHub:
https://github.com/kubernetes/kubernetes/pull/12599/files#r39430697